pkg/rollout/traffic.go

```diff
@@ -0,0 +1,184 @@
+package rollout
+
+import (
+	"time"
+
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/health"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/run/v1"
+)
+
+// trafficBasedOnDiagnosis updates a service's traffic configuration
+// based on the diagnosis.
+// If traffic should not changed, nil is returned.
+func (r *Rollout) trafficBasedOnDiagnosis(svc *run.Service, diagnosis health.DiagnosisResult, stable, candidate string) ([]*run.TrafficTarget, error) {
+	switch diagnosis {
+	case health.Inconclusive:
+		r.log.Debug("health check inconclusive")
+		return nil, nil
+	case health.Healthy:
+		r.log.Debug("healthy candidate")
+		lastRollout := svc.Metadata.Annotations[LastRolloutAnnotation]
+		enoughTime, err := r.hasEnoughTimeElapsed(lastRollout, r.strategy.TimeBetweenRollouts)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not determine if roll out is allowed")
+		}
+		if !enoughTime {
+			r.log.WithField("lastRollout", lastRollout).Debug("no enough time elapsed since last roll out")
+			return nil, nil
+		}
+		r.log.Debug("rolling forward")
+		return r.RollForwardTraffic(svc.Spec.Traffic, stable, candidate), nil
+	case health.Unhealthy:
+		r.log.Info("unhealthy candidate, rollback")
+		r.shouldRollback = true
+		return r.RollbackTraffic(svc.Spec.Traffic, stable, candidate), nil
+	default:
+		return nil, errors.Errorf("invalid candidate's health diagnosis %v", diagnosis)
+	}
+}
+
+// RollForwardTraffic updates the traffic configuration to increase the traffic
+// to the candidate.
+//
+// It creates new traffic configurations for the candidate and stable revisions
+// and respects user-defined revision tags.
+func (r *Rollout) RollForwardTraffic(traffic []*run.TrafficTarget, stable, candidate string) []*run.TrafficTarget {
```

Comment:
these probably should be unexported since the only place they're called from are unexported.